### PR TITLE
Add harmonize collections tool (or whatever other name)

### PR DESCRIPTION
**Pull request by lldelisle**
Add a new tool to synchronize collections. After discussion with IWC it seemed needed for workflows.

## How to test the changes?
(Select all options that apply)
- ?? I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- partially This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  In theory the tests in the tool wrapper should be sufficient, no?
## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by lldelisle** [comment]
@wm75 if you find other tests to include, please tell me.

File: lib/galaxy/tools/__init__.py [hunk 1]
```diff
@@ -3563,6 +3563,65 @@ def produce_outputs(self, trans, out_data, output_collections, incoming, history
...
+        # Get the list of final identifiers
+        final_sorted_identifiers = [element.element_identifier for element in elements1
+                                    if element.element_identifier in old_elements2_dict]
+        # Raise Exception if it is empty
+        if len(final_sorted_identifiers) == 0:
+            hdca1_history_name = f"{hdca1.hid}: {hdca1.name}"
+            hdca2_history_name = f"{hdca2.hid}: {hdca2.name}"
+            message = f"The two collections '{hdca1_history_name}' and '{hdca2_history_name}'" \
+                " have no identifier in common. It is not possible to synchronize them."
+            raise Exception(message)
```
Comment (mvdbeek) [inline lib/galaxy/tools/__init__.py]: I think that's not necessarily an error, you may just legitimately have an empty collection

Comment (lldelisle) [reply lib/galaxy/tools/__init__.py]: OK

File: lib/galaxy/tools/synchronize_two_collections_list.xml [hunk 2]
```diff
@@ -0,0 +1,196 @@
+<tool id="__SYNCHRONIZELISTS__"
```
Comment (mvdbeek) [inline lib/galaxy/tools/synchronize_two_collections_list.xml]: Can you add this to https://github.com/galaxyproject/galaxy/blob/dev/test/functional/tools/sample_tool_conf.xml#L294 and https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/sample/tool_conf.xml.sample#L45 ?

With that you can should be able to run the tool test with `./run_tests.sh -framework -id __SYNCHRONIZELISTS__`

**Comment by jmchilton** [comment]
Amazing idea for a tool - I wish I had thought of this - definitely we need to have this. I don't love the verb synchronize (https://www.thesaurus.com/browse/synchronize) for this... it feels like you're making the collection equal instead of matching them up? I think harmonize (https://www.thesaurus.com/browse/harmonize) feels better or maybe correlate or coordinate. This is not a hill I'm going to die on though - if y'all think synchronize is better I'm fine with it.

**Comment by mvdbeek** [comment]
😆 we've had a 5 minute brainstorm on what to call it among all non-native speakers ... I like harmonize.

**Comment by lldelisle** [comment]
I told you guys, naming is the most difficult part...

**Comment by bernt-matthias** [comment]
I'm joining the name game. Formally you compute am intersection of the identifiers. Is intersection to formal?

**Comment by jdavcs** [comment]
> I'm joining the name game. Formally you compute am intersection of the identifiers. Is intersection to formal?

I had to read the example in the help section to understand what harmonize means in this context. Intersect/intersection, on the other hand, is immediately clear. Not too formal, IMO.

**Comment by jmchilton** [comment]
> There are 2 hard problems in computer science: cache invalidation, naming things, and off-by-1 errors.

My immediate intuition for what an intersecting collection tool would do... would be to produce the set intersection of the datasets. A totally reasonable operation - maybe you filter the collection two different ways using existing tools and then you want to insect the results. I will admit I don't know what harmonize would do right away but... at least it doesn't lead me down the wrong path?

> "I need to do a thing... I'll review the tools... this isn't intersection... what is harmonize... I don't know but I'll look more closely I guess it might be that."

Maybe the approach here is... to add three tools a "Union Datasets" (CxC -> C), a "Intersect Datasets" (CxC -> C), and an "Intersect Identifiers" tool (CxC -> CxC). 

I wonder if this is a normal thing in another, bigger map reduce framework or in that spreadsheet software from EvilCorpTM.

File: lib/galaxy/tools/__init__.py [hunk 3]
```diff
@@ -3563,6 +3563,67 @@ def produce_outputs(self, trans, out_data, output_collections, incoming, history
...
+                                    if element.element_identifier in old_elements2_dict]
+        # Raise Exception if it is empty
+        if len(final_sorted_identifiers) == 0:
+            # Create empty collections:
+            output_collections.create_collection(
+                next(iter(self.outputs.values())), "output1", elements={}, propagate_hda_tags=False
+            )
+            output_collections.create_collection(
+                next(iter(self.outputs.values())), "output2", elements={}, propagate_hda_tags=False
+            )
```
Comment (jdavcs) [inline lib/galaxy/tools/__init__.py:3597]: Now that an error is not raised, I suppose we need an early return here?

File: lib/galaxy/tools/__init__.py [hunk 4]
```diff
@@ -3563,6 +3563,67 @@ def produce_outputs(self, trans, out_data, output_collections, incoming, history
...
+        # Create outputs:
+        new_elements1 = {}
+        new_elements2 = {}
+        for identifier in final_sorted_identifiers:
+            # First collection output
```
Comment (jdavcs) [inline lib/galaxy/tools/__init__.py]: The logic for first and second collection output generation can be factored out into a simple helper method: the only difference in lines 3602-3607 and 3609-3614 is the 2 dictionaries (old and new elements) - you could just pass them as arguments and call the helper twice.

File: lib/galaxy/tools/__init__.py (hunk 3 above)
Comment (lldelisle) [reply lib/galaxy/tools/__init__.py:3597]: Yes, thanks

File: lib/galaxy/tools/__init__.py (hunk 4 above)
Comment (lldelisle) [reply lib/galaxy/tools/__init__.py]: thanks for the tip

**Comment by lldelisle** [comment]
> Whichever name is selected for the tool, can we please have the tool filename updated as well?

Good point...

**Comment by lldelisle** [comment]
I let you debate on the name.
For info, the idea of this tool arises for the need to reorder 2 collections so we can compute operations between them and make sure they are 'in phase'. So the initial goal of this tool was to reorder.

**Comment by lldelisle** [comment]
My tests are locally passing. Ping me if I need to do something. Feel free to rename the tool the way you feel is better. If you prefer "intersect identifiers of collections" it is fine.

**Comment by jdavcs** [comment]
The 2 selenium failing tests are unrelated. I've restarted the rest (and they've passed!)

Name: "harmonize" is perfectly fine. Besides, I think @jmchilton made an excellent point about the "intersection" alternative (..."I don't know what harmonize would do right away but... at least it doesn't lead me down the wrong path"). But the discussion was very helpful (also, @lldelisle thanks for the added context!).

**Review by jdavcs — APPROVED** [review]
Thank you, @lldelisle - this is very neat! Also, thank you for addressing all the comments!

**Comment by lldelisle** [comment]
You're welcome. It was fast.

**Comment by lldelisle** [comment]
Should I rebase it to 23.1 so people can use it earlier?

**Comment by jdavcs** [comment]
> Should I rebase it to 23.1 so people can use it earlier?

Unfortunately no: once we reach the freeze step in the release process, we can only add bug fixes to that release, which is the case with 23.1.

**Pull request merged**